File: templates/voice.py
import openai
import pyttsx3
import speech_recognition as sr
import time

#set tour openai api key
openai.api_key=""

#initialize the text-to-speech engine
engine = pyttsx3.init()

def transcribe_audio_to_text(filename):
    recognizer =sr.Recognizer()
    with sr.AudioFile(filename) as source:
        audio=recognizer.record(source)
    try:
        return recognizer.recognize_google(audio)
    except:
        print("Skipping unknown error")

def generate_response(prompt):
    response = openai.Completion.create(
        engine="text-davinci-003",
        prompt=prompt,
        max_tokens=4000,
        n=1,
        stop=None,
        temperature=0.5,
    )
    return response["choices"][0]["text"]

def speak_text(text):
    engine.say(text)
    engine.runAndWait()


def main():
    while True:
        # Wait for user to say "Genius"
        print("Say 'Genius' to start recording your question...")
        with sr.Microphone() as source:
            recognizer = sr.Recognizer()
            audio = recognizer.listen(source)
            try:
                transcription =  recognizer.recognize_google(audio)
                if transcription.lower()=="genius":
                    #Record Audio
                    filename = "input.wav"
                    print("Say youe question...")
                    with sr.Microphone() as source:
                        recognizer = sr.Recognizer()
                        source.pause_threshold=1
                        audio = recognizer.listen(source,phrase_time_limit=None, timeout=None)
                        with open(filename,"wb")as f:
                            f.write(audio.get_wav_data())

                    # Transcribe audio to text
                    text = transcribe_audio_to_text(filename)
                    if text:
                        print(f"You said: {text}")

                        #Generate response using GPT-3
                        response = generate_response(text)
                        print("GPT-3 says: {response}")

                        #Read response using text-to-speech
                        speak_text(response)
            except Exception as e:
                print("An Error occured : {}",format(e))

if __name__ == "__main__":
    main()

File: static/js/voice.js
import { createReadStream, writeFile } from 'fs';
import { createInterface } from 'readline';
import { exec } from 'child_process';
import { AudioContext } from 'web-audio-api';

// Simplified audio recording and storage function
async function recordAudio(filename) {
  return new Promise((resolve, reject) => {
    // Record audio and save it as the specified file
    exec(`arecord -f S16_LE -r 16000 -d 5 ${filename}`, (error) => {
      if (error) {
        console.error(`Recording error: ${error}`);
        reject(error);
      } else {
        console.log(`Audio recorded and saved to ${filename}`);
        resolve(createReadStream(filename));
      }
    });
  });
}

async function main() {
  const rl = createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  while (true) {
    console.log("Say 'kitchen' to start recording your question...");
    await recordAudio('input.wav'); // Record initial keyword trigger audio

    // Process the audio file to get the transcription (This function needs to be defined)
    const transcription = await transcribeAudioToText('input.wav');
    if (transcription.toLowerCase() === 'kitchen') {
      console.log('Say your question...');
      await recordAudio('question.wav'); // Record the question audio

      // Transcribe and handle the question audio
      const questionText = await transcribeAudioToText('question.wav');
      if (questionText) {
        console.log(`You said: ${questionText}`);

        // Generate a response using GPT-3 (This function needs to be defined)
        const response = await generateResponse(questionText);
        console.log(`GPT-3 says: ${response}`);

        // Read the response using text-to-speech (This function needs to be defined)
        speakText(response);
      }
    }
  }
}

// Ensure the main function handles errors gracefully
main().catch((err) => console.error('An Error occurred:', err));
